src/ott/neural/models/base_solver.py

```diff
@@ -0,0 +1,485 @@
+# Copyright OTT-JAX
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import abc
+import pathlib
+import types
+from typing import Any, Callable, Dict, Literal, Mapping, Optional, Tuple, Union
+
+import jax
+import jax.numpy as jnp
+
+import optax
+from flax.training import train_state
+
+from ott.geometry import costs, pointcloud
+from ott.problems.linear import linear_problem
+from ott.problems.quadratic import quadratic_problem
+from ott.solvers.linear import sinkhorn
+
+__all__ = ["BaseNeuralSolver", "ResampleMixin", "UnbalancednessMixin"]
+
+
+class BaseNeuralSolver(abc.ABC):
+  """Base class for neural solvers.
+
+  Args:
+    iterations: Number of iterations to train for.
+    valid_freq: Frequency at which to run validation.
+  """
+
+  def __init__(self, iterations: int, valid_freq: int, **_: Any):
+    self.iterations = iterations
+    self.valid_freq = valid_freq
+
+  @abc.abstractmethod
+  def setup(self, *args: Any, **kwargs: Any):
+    """Setup the model."""
+
+  @abc.abstractmethod
+  def __call__(self, *args: Any, **kwargs: Any):
+    """Train the model."""
+
+  @abc.abstractmethod
+  def transport(self, *args: Any, forward: bool, **kwargs: Any) -> Any:
+    """Transport."""
+
+  @abc.abstractmethod
+  def save(self, path: pathlib.Path):
+    """Save the model."""
+
+  @abc.abstractmethod
+  def load(self, path: pathlib.Path):
+    """Load the model."""
+
+  @property
+  @abc.abstractmethod
+  def training_logs(self) -> Dict[str, Any]:
+    """Return the training logs."""
+
+
+class ResampleMixin:
```

Comment:
Let's prefer composition over inheritance, i.e., I would remove the mixins and pass these objects the the class during initialization (it's better when someone wants to implement try different combinations).

Reply:
So you would instantiate this class, and hence all the classes should have a `Resampler` as an attribute?